cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java

```diff
@@ -383,22 +412,82 @@ public void apply() throws Exception {
     }
   }
 
+  // Returns true if the upgrade flag is set. Upgrade could have completed earlier than this since this flag is
+  // updated asynchronously.
+  public boolean isUpgradeComplete() {
+    return appVerUpgrade.get();
+  }
+
   /**
-   * Method to add version in JobKey and VersionKey in SchedulerStore.
+   * Method to add version to row key in SchedulerStore.
    *
-   * @throws Exception
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws DatasetManagementException
    */
-  public void upgrade() throws Exception {
-    // Create table instance.
-    initializeScheduleTable();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() throws Exception {
-          upgradeJobs();
-          upgradeTriggers();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
+    try {
+      factory.createExecutor(ImmutableList.of((TransactionAware) table))
+        .execute(new TransactionExecutor.Subroutine() {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
```

Comment:
why not compare result with `Bytes.toBytes(ProjectInfo.getVersion().toString())`?